src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardStream.cs

```diff
@@ -0,0 +1,382 @@
+                                }
+                                catch
+                                {
+                                    // ignore exception
+                                }
+                            }
+                        }
+                        finally
+                        {
+                            _nativeWrapper.Dispose();
+                            _arrayPool.Return(_streamWriteHelper.CompressedBufferInfo.Bytes);
+                        }
+                        break;
+                    case CompressionMode.Decompress:
+                        _nativeWrapper.Dispose();
+                        _arrayPool.Return(_streamReadHelper.CompressedBufferInfo.Bytes);
+                        break;
+                }
+
+                _disposed = true;
+            }
+        }
+
+        public override void Flush()
+        {
+            foreach (var outputBufferPosition in _nativeWrapper.StepwiseFlush(_streamWriteHelper.CompressedBufferInfo))
+            {
+                _streamWriteHelper.WriteBufferToCompressedStream(count: outputBufferPosition);
+            }
+            _compressedStream.Flush();
+            _flushed = true;
+        }
+
+        public override int Read(byte[] outputBytes, int outputOffset, int count) // Decompress
+        {
+            if (!CanRead) throw new InvalidDataException("Read is not accessible.");
+
+            var uncompressedOutputBuffer = new BufferInfo(outputBytes, outputOffset);
+
+            using (var operationContext = _nativeWrapper.InitializeOperationContext(
+                _streamReadHelper.CompressedBufferInfo,
+                uncompressedOutputBuffer)) // operation result
+            {
+                var length = 0; // the result size
+                var remainingCount = count;
+
+                while (remainingCount > 0)
+                {
+                    var currentDataPosition = _streamReadHelper.ReadCompressedStreamToBufferIfAvailable(
+                        recommendedCompressedSize: _nativeWrapper.RecommendedInputSize,
+                        out int remainingCompressedBufferSize);
+
+                    // decompress input to output
+                    _nativeWrapper.Decompress(
+                        operationContext,
+                        compressedOffset: currentDataPosition,
+                        compressedSize: remainingCompressedBufferSize,
+                        uncompressedSize: remainingCount,
+                        out int compressedBufferPosition,
+                        out int uncompressedBufferPosition);
+
+                    if (!_streamReadHelper.TryPrepareDataForNextAttempt(
+                        compressedBufferPosition,
+                        uncompressedBufferPosition))
+                    {
+                        break;
+                    }
+
+                    length += uncompressedBufferPosition;
+                    remainingCount -= uncompressedBufferPosition;
```

Comment:
This still looks wrong... I don't see how you can add a position to a length. (It's like mixing units).

Reply:
`length` - is "count" of bytes which were processed(deserialized) during the current Read call (which is obvious).
`uncompressedBufferPosition` - is the number of uncompressed bytes in the current attempt (a loop iteration). Because a position before each attempt is set to 0. 

So, adding uncompressedBufferPosition on each iteration to length is just updating the length value on the number of bytes that were processed in the current loop iteration.